### PR TITLE
Add free convection example

### PR DESCRIPTION
**Pull request by Raamkrishnan**
### IBAMR Pull Request Checklist
- [ ] Does the test suite pass?
- [ ] Was clang-format run?
- [ ] Were relevant issues cited? Please remember to add `Fixes #12345` to close
      the issue automatically if we are fixing the problem.
- [ ] Is this a change others will want to know about? If so, then has a
      changelog entry been added?
- [ ] If new data structures have been added to a class then have they been set
      up appropriately for restarts? If so, ensure that the restart version
      number is incremented.
- [ ] Does this change include a bug fix or new functionality? If so, a new test
      or tests should be added. New tests should run quickly (less than a minute
      in release mode). If possible, an older test should gain a new option so
      that we do not need to compile more test executables.
- [ ] Did you (if your account has permission to do so) set relevant labels on
      GitHub for the pull request?

In this example, steady natural convection in a concentric annulus is simulated. ```BrinkmanAdvDiffBcHelper``` class is used to impose flux boundary conditions on the cylinders and ```BrinkmanPenalizationRigidBodyDynamics``` class is used to impose no-slip on the cylinder surface. In this example, we solve both momentum and energy equations in a non-dimensional form. For now, I have added this example in ```examples/adv_diff/ex8```. What is the right directory to place this example? IB, IBLevelSet, or navier_stokes? 

More details of this example are given in Section 5.7 of "Thirumalaisamy, R., Patankar, N.A. and Bhalla, A.P.S., 2022. Handling Neumann and Robin boundary conditions in a fictitious domain volume penalization framework. Journal of Computational Physics, 448, p.110726."

### CODE REVIEW

File: examples/adv_diff/ex8/BoussinesqForcing.cpp [hunk 1]
```diff
@@ -0,0 +1,203 @@
...
+// license. The full text of the license can be found in the file
+// COPYRIGHT at the top level directory of IBAMR.
+//
+// ---------------------------------------------------------------------
+
+#include "BoussinesqForcing.h"
+
+/////////////////////////////// INCLUDES /////////////////////////////////////
+
+#include <SAMRAI_config.h>
```
Comment (drwells) [inline examples/adv_diff/ex8/BoussinesqForcing.cpp:18]:
```suggestion
```
Not super important but I'm trying to get rid of all of these. SAMRAI headers always include this header.

File: examples/adv_diff/ex8/LevelSetInitialCondition.cpp [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+// license. The full text of the license can be found in the file
+// COPYRIGHT at the top level directory of IBAMR.
+//
+// ---------------------------------------------------------------------
+
+#include "LevelSetInitialCondition.h"
+
+/////////////////////////////// INCLUDES /////////////////////////////////////
+#include <SAMRAI_config.h>
+
```
Comment (drwells) [inline examples/adv_diff/ex8/LevelSetInitialCondition.cpp:18]:
```suggestion
```

File: examples/adv_diff/ex8/write_ascii.cpp [hunk 3]
```diff
@@ -0,0 +1,38 @@
+#include <mpi.h>
+
+#include <cassert>
+#include <fstream>
+#include <iostream>
+#include <vector>
+int
+main(int argc, char** argv)
```
Comment (drwells) [inline examples/adv_diff/ex8/write_ascii.cpp:8]: Could you add a comment block before `main` explaining what this file does? I'm lost in all the MPI file commands.

File: examples/adv_diff/ex8/example.cpp [hunk 4]
```diff
@@ -0,0 +1,966 @@
...
+//
+// IBAMR is free software and is distributed under the 3-clause BSD
+// license. The full text of the license can be found in the file
+// COPYRIGHT at the top level directory of IBAMR.
+//
+// ---------------------------------------------------------------------
+
+// Config files
+#include <SAMRAI_config.h>
+
```
Comment (drwells) [inline examples/adv_diff/ex8/example.cpp:16]:
```suggestion
```

**Review by drwells — CHANGES_REQUESTED** [review]
I don't know enough about the topic to give a really thorough review - it's probably fine if you and Amneet are happy with it. I have a few cosmetic requests.

File: examples/adv_diff/ex8/write_ascii.cpp (hunk 3 above)
Comment (Raamkrishnan) [reply examples/adv_diff/ex8/write_ascii.cpp:13]: This file is used to read the MPI binary file written by ```example.cpp``` and output it in ASCII file format for postprocessing. I am not good at MPI. If there is a way for me to write the data in ASCII format directly in ```example.cpp```, I am ready to do that. I just followed the code from ```tests/IIM/taylor_couette_2d.cpp``` which writes velocity profile:

https://github.com/IBAMR/IBAMR/blob/d5981eb355272b92c655006616d4c9edaa5a2cae/tests/IIM/taylor_couette_2d.cpp#L991


I used this code snippet for the temperature profile.

**Comment by Raamkrishnan** [comment]
> I don't know enough about the topic to give a really thorough review - it's probably fine if you and Amneet are happy with it. I have a few cosmetic requests.

We already have 5 examples on this topic: ```examples/adv_diff/ex3``` - ```examples/adv_diff/ex7```. The main difference is the current example solves the Navier-Stokes as well along with the advection-diffusion equation. I am not sure where this example should be placed. @amneetb Your suggestion?

**Comment by amneetb** [comment]
I am OK with this example being here. Is there a README file that gives a short description about the example?

File: examples/adv_diff/ex8/README [hunk 5]
```diff
@@ -0,0 +1,3 @@
+This example simulates steady natural convection in a concentric annulus. The boundary conditions on the cylinders are imposed through the flux-based volume penalization method. The flux-forcing function is constructed numerically. Please refer Section 5.7 of "Thirumalaisamy, R., Patankar, N.A. and Bhalla, A.P.S., 2022, Handling Neumann and Robin boundary conditions in a fictitious domain volume penalization framework, Journal of Computational Physics, 448, p.110726" for more details on this example.
```
Comment (Raamkrishnan) [inline examples/adv_diff/ex8/README:1]: @amneetb Yes. There is a README file.

**Review by amneetb — APPROVED** [review]

**Review by drwells — APPROVED** [review]

**Pull request merged**